Show task and activity notifications in the Inbox panel

octaedro opened the pull request:
Fixes #6737

This PR adds abbreviated cards to show task and activity notifications in the Inbox panel.

The original design changed a bit, so take the image in [this comment](https://github.com/woocommerce/woocommerce-admin/issues/6737#issuecomment-840831948) as a reference. 

Also, the Inbox panel width will be 80px smaller.

### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

-   [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
-   [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

![screenshot-one wordpress test-2021 05 21-09_50_58](https://user-images.githubusercontent.com/1314156/119140942-1ab6c100-ba1b-11eb-889a-c284378bf0fa.png)

### Detailed test instructions:

-   Go to the `Orders` page (the important thing here is to go to any page other than `Home`).
-   Open the `Inbox` panel.
-   Verify the `Orders`, `Reviews`, and `Stock` abbreviated cards are not visible when the Setup tasks are visible in the `Home`screen.
-   Verify that the card `Things to do next` is only visible when the extended tasks are visible.
-   ~The `Store Alerts` are indicated as critical tasks (`# critical alerts`). To add a store alert (a critical task to show), you can use the plugin [wc-admin-test-helper](https://github.com/woocommerce/woocommerce-admin-test-helper) or change the type to `update` of an (unactioned and not deleted) admin note. You can use a SQL statement like this:~

```
UPDATE `wp_wc_admin_notes` SET `type` = 'update', `is_deleted` = 0, `status` = 'unactioned' WHERE `wp_wc_admin_notes`.`name` = 'wc-admin-onboarding-payments-reminder'
```

To return the note back to normal, you should execute a statement like:

```
UPDATE `wp_wc_admin_notes` SET `type` = 'info', `is_deleted` = 0, `status` = 'unactioned' WHERE `wp_wc_admin_notes`.`name` = 'wc-admin-onboarding-payments-reminder'
```

-   Click on every abbreviated card and verify the user is redirected to the `Home` screen and the selected panel is opened. In the case of `Things to do next`, the user should be redirected to the Home screen with the extended task list anchored to the top of the view.
-   Verify that the "unread indicator" in the Inbox Icon is active when there is an unread note or an abbreviated panel to show.
-   Enable debug messages inside browser devtools and verify the events are being recorded correctly. The event `wcadmin_activity_panel_click` should be recorded after clicking with the selected task as a prop.


### UPDATE ###
After chatting with @elizaan36 , the `critical alerts` tag is removed for now (commit cf65b63).

<!-- 
Please add your test instructions to `/TESTING-INSTRUCTIONS.md`.
-->

<!--- Please add a Changelog note

Enter a changelog note following the WooCommerce core format using prefixes of Enhancement:, Tweak:, Dev:, Fix:, Performance: to readme.txt under the "unreleased" list at the top of the changelog. If none exists, please add it. Also include PR number.

If changes pertain to a package, also update CHANGELOG.md in the package's folder in a similar manner.--->

octaedro committed:
Added abbreviated panels

This commit adds abbreviated panels

octaedro committed:
Added notifications getter

octaedro committed:
Variables renamed

octaedro committed:
Added unread-indicators refactor

octaedro committed:
Open panel by default

octaedro committed:
Refactor unread-indicators

octaedro committed:
Renamed a few files and added event recording

octaedro committed:
Modified "critical alert" presentation

octaedro committed:
Removed useless control

octaedro committed:
Renamed const

octaedro committed:
Added control to InboxPanel component

octaedro committed:
Multiple critical alerts handling

octaedro committed:
Fixed styles

octaedro committed:
Moved Inbox panel styles

# Conflicts:
#	packages/experimental/src/inbox-note/style.scss

octaedro committed:
Added tests

octaedro committed:
Inbox panel width reduced

octaedro committed:
Small refactor for unread notifications

octaedro committed:
Renamed abbreviated card component

octaedro committed:
Added changelog

# Conflicts:
#	readme.txt

# Conflicts:
#	readme.txt

octaedro committed:
Renamed inbox-panel and the cards config file

octaedro committed:
Renamed unread notifications variable

octaedro committed:
Fixed abbreviated card box-shadow

octaedro committed:
Small refactor to unread-indicators file

octaedro committed:
Refactored method getInitialState

octaedro committed:
Added scroll to task list

# Conflicts:
#	client/task-list/task-list.js

octaedro committed:
Small CSS changes to titles

jeffstieler commented:
In testing this, "things to do" didn't show up as expected in the inbox panel:

![Screen Shot 2021-05-24 at 3 10 44 PM](https://user-images.githubusercontent.com/63922/119396226-9a09f600-bca2-11eb-8e24-40c7c3bfaa15.png)

![Screen Shot 2021-05-24 at 3 11 29 PM](https://user-images.githubusercontent.com/63922/119396250-9eceaa00-bca2-11eb-8d8c-072f57a64520.png)

But did show when there was a store alert:

![Screen Shot 2021-05-24 at 3 11 11 PM](https://user-images.githubusercontent.com/63922/119396262-a2fac780-bca2-11eb-93f6-6c20d5763328.png)

Which felt weird to me because the "critical task" referenced is the visible coupon alert, but clicking the "things to do next" took me to the homescreen with the task list focused, where there was nothing critical.

jeffstieler commented on readme.txt:
```diff
@@ -126,6 +126,7 @@ Release and roadmap notes are available on the [WooCommerce Developers Blog](htt
 - Update: Adding setup required icon for non-configured payment methods #6811
 - Update: Task list component with new Experimental Task list. #6849
 - Update: Redirect to WC Home after setting up a payment method #6891
+- Dev: Show task and activity notifications in the Inbox panel #7017
```
Why `Dev`? Shouldn't this be `Add` or `Update`?

octaedro replied:
You're right, `Add` sounds good to me. This was added in the commit b788750

jeffstieler commented on client/header/activity-panel/unread-indicators.js:
```diff
@@ -57,3 +72,71 @@ export function getUnreadNotes( select ) {
+	const thingsToDoNext = applyFilters(
+		'woocommerce_admin_onboarding_task_list',
+		[]
+	);
```
Is there another way to get these tasks? Applying this filter in multiple spots is a code smell to me. If it's not possible any other way, we need to at least pass in the same parameters to the filter (like [`query`](https://github.com/woocommerce/woocommerce-admin/blob/c863e5d5c508d3de023df42a44e95557d5297890/client/task-list/tasks.js#L262)).

Also it seems that perhaps plugins aren't being loaded on "embed" pages (like WooCommerce > Orders) .. so 3PD code that would filter in a task isn't running (see https://github.com/woocommerce/woocommerce-admin/pull/7017#issuecomment-847272797).

octaedro replied:
> Is there another way to get these tasks?

I'm using the same way we use to get the extended tasks in `Things to do next`. Sending the same parameters (`query`) to the filter is a good idea. I added them in the commit c1ef806

> Also it seems that perhaps plugins aren't being loaded on "embed" pages (like WooCommerce > Orders) .. so 3PD code that would filter in a task isn't running

This makes sense. Even [in our example](https://github.com/woocommerce/woocommerce-admin/blob/main/docs/examples/extensions/add-task/woocommerce-admin-add-task-example.php#L17) that's being prevented.
I think that until we switch the way we handle the extended tasks to `SlotFill` we will continue having this kind of issue.

jeffstieler commented on client/header/activity-panel/unread-indicators.js:
```diff
@@ -57,3 +72,71 @@ export function getUnreadNotes( select ) {
+	];
+
+	return applyFilters(
+		'woocommerce_admin_abbreviated_notifications',
```
I seem to have gotten the idea in my head that we didn't want to continue using the pattern of "filtered array of data that turns into components". Should we be using SlotFill here instead? cc: @samueljseay _(note: yes, I continued the components-as-data pattern [here](https://github.com/Automattic/woocommerce-payments/pull/1799/files#diff-062eb1e5dc64b4146f2536ba4987039bedb128b2d938de7217119a9beaa2fd87R10), but those components need to slot in to our existing task list filter of data-represented-components)_ 😅

samueljseay replied:
@octaedro sorry if the messaging hasn't been too clear on this, but yeah even aside from "filtered array of data that turns into components" not being a great pattern there are too many gotchas with JS filters that make them really hard to use if you're trying to use them as an extensibility point. Since that really only leaves us with SlotFill as an alternative I think we need to go that route here. 

I realise that SlotFill can take some time to get your head around so please do reach out to @jeffstieler, @louwie17 or even @psealock or anyone from Gezora because they've all had experience implementing this pattern.

jeffstieler replied:
Perhaps the easiest path forward would be to remove the extensibility from this PR and work on it in a follow up?

octaedro replied:
Thank you @jeffstieler and @samueljseay, you're right, using SlotFill for the abbreviated panels is a better idea. If it's ok with you I'll remove the extensibility and work on a SlotFill approach in a follow-up PR.
The extensibility was removed in the commit fbb3aaa

jeffstieler commented on client/header/activity-panel/panels/inbox/inbox-panel.js:
```diff
@@ -0,0 +1,48 @@
+						return (
+							<AbbreviatedCard
+								content={ sprintf( content, count ) }
+								critical={ critical ?? 0 }
```
Does this need to be a prop? Can it just be part of the content? This doesn't feel like a generic property to me since (right now) AFAICT it's only for the "things to do next" item and the label isn't customizable (it always says "X critical alerts").

octaedro replied:
I really like the idea! That'd make the component more generic and simple 👍 
This change was added in the commit 373e19e

jeffstieler commented on client/header/activity-panel/panels/inbox/cards.js:
```diff
@@ -0,0 +1,53 @@
+const REVIEWS_PANEL_ID = 'reviews-panel';
+const STOCK_PANEL_ID = 'stock-panel';
+
+export const cards = applyFilters( 'woocommerce_admin_abbreviated_card_list', [
```
See comment below about filters vs. SlotFill.

octaedro replied:
I removed this filter in the commit 9c729f4

jeffstieler commented on client/header/activity-panel/panels/inbox/abbreviated-card.js:
```diff
@@ -0,0 +1,59 @@
+
+export const AbbreviatedCard = ( {
+	content,
+	critical,
```
Does this need to be a prop? Can it just be part of the content? This doesn't feel like a generic property to me since (right now) AFAICT it's only for the "things to do next" item and the label isn't customizable (it always says "X critical alerts").

jeffstieler commented on client/header/activity-panel/panels/inbox/abbreviated-card.js:
```diff
@@ -0,0 +1,59 @@
+import { Link } from '@woocommerce/components';
+import { recordEvent } from '@woocommerce/tracks';
+
+export const AbbreviatedCard = ( {
```
Can this be an exported/public component? I think design has been looking at our UI elements as "available building blocks" for future work, but they need part of our published packages to be usable by 3PDs and other teams.

octaedro replied:
I like the idea. I moved the component in the commit 6578cc3

jeffstieler commented on client/header/activity-panel/index.js:
```diff
@@ -39,7 +39,7 @@ const HelpPanel = lazy( () =>
 const InboxPanel = lazy( () =>
 	import(
-		/* webpackChunkName: "activity-panels-inbox" */ '../../inbox-panel'
+		/* webpackChunkName: "activity-panels-help" */ './panels/inbox/inbox-panel'
```
Why change the chunk name here?

octaedro replied:
Nice catch! I modified it by mistake. It was fixed in the commit fe250be

octaedro committed:
Fixed changelog

# Conflicts:
#	readme.txt

# Conflicts:
#	readme.txt

octaedro committed:
Added param to filter `woocommerce_admin_onboarding_task_list`

octaedro committed:
Removed extensibility from `getAbbreviatedNotifications`

octaedro committed:
Fixed chunk name

octaedro committed:
Removed `critical` prop from `AbbreviatedCard` comopnent

octaedro committed:
Moved AbbreviatedCard component to `packages`

This commit moves the component `AbbreviatedCard` to `packages`

# Conflicts:
#	docs/components/_sidebar.md
#	packages/components/CHANGELOG.md
#	packages/components/src/index.js

# Conflicts:
#	packages/components/CHANGELOG.md

octaedro committed:
Removed `critical alerts` tag from abbreviated card

This commit removes the tag `critical alerts` from the `Things to do next` abbreviated card

octaedro commented:
Hi @jeffstieler, than you for your review.
I applied the changes you suggested. Could you give this PR another look?

octaedro committed:
Removed filter `woocommerce_admin_abbreviated_card_list`

octaedro committed:
Fixed icon

jeffstieler commented:
When trying to test the "things to do next" I'm getting a JS error: `Uncaught TypeError: Cannot read property 'hasUnreadNotifications' of undefined` using the "Add Task" example plugin.

octaedro committed:
Added defaut value to `hasUnreadNotifications`

octaedro commented:
Hi @jeffstieler,
Thank you for the review and sorry for that error.
I tried to reproduce it but I couldn't, but I think that adding a default value to `hasUnreadNotifications` would solve the problem. I did it in the commit 730afc2
If you use the `Add task` example plugin, [this line](https://github.com/woocommerce/woocommerce-admin/blob/main/docs/examples/extensions/add-task/woocommerce-admin-add-task-example.php#L17) will prevent it from working on pages like: `products` or `orders`, but it will still work on any `Analytics` pages.

jeffstieler committed:
Fix mapSelect error when the dismissed tasks option isn't populated.

jeffstieler commented:
> I tried to reproduce it but I couldn't, but I think that adding a default value to hasUnreadNotifications would solve the problem

Sorry for the lack of detail in describing this error. It was stemming the combination of having no `woocommerce_task_list_dismissed_tasks` option in the database and having extension tasks filtered in. This is the fix: a82003f

jeffstieler commented:
Here are a few minor things found in testing (going to review code now):

We should handle singular/plural cases in the strings:
<img width="353" alt="Screen Shot 2021-06-02 at 8 47 41 AM" src="https://user-images.githubusercontent.com/63922/120483025-f3db8180-c36e-11eb-9af8-d140aa0fc158.png">

And when clicking on "things to do next" when the setup task list is complete/hidden, scrolling to the extended task list seems to overshoot it:
<img width="696" alt="Screen Shot 2021-06-02 at 8 49 28 AM" src="https://user-images.githubusercontent.com/63922/120483176-14a3d700-c36f-11eb-8b94-e69578bfae89.png">

jeffstieler commented on client/header/activity-panel/panels/inbox/cards.js:
```diff
@@ -0,0 +1,52 @@
+		icon: <Bell />,
+		title: __( 'Things to do next', 'woocommerce-admin' ),
+		/* translators: Things the user has to do */
+		content: __( 'You have %s new things to do', 'woocommerce-admin' ),
```
We'll need to handle the singular/plural cases here.

octaedro replied:
Good eyes! This was added in the commit e1000b4

jeffstieler commented on client/header/activity-panel/panels/inbox/inbox-panel.js:
```diff
@@ -0,0 +1,58 @@
+ */
+import './inbox.scss';
+import NotesPanel from '../../../../inbox-panel';
+import { cards } from './cards';
```
Do we need to represent the cards as data? Could we wire this component up to state instead and have it decide which `AbbreviatedCard`s to render explicitly? I'm just not seeing the benefit to representing the cards and unread counts as objects that have to be matched up later in `getCardByName()`, since we aren't applying a filter or anything here.

octaedro replied:
Good idea! I changed it in the commit 8a9bdac and added tests in the commit fa25dc3

jeffstieler commented on client/header/activity-panel/unread-indicators.js:
```diff
@@ -57,3 +65,65 @@ export function getUnreadNotes( select ) {
+		{
+			name: 'stockNotices',
+			count: getLowStockProducts( select ),
+		},
```
See other comment about moving this into the `InboxPanel` component.

jeffstieler commented on client/homescreen/activity-panel/index.js:
```diff
@@ -69,6 +70,13 @@ export const ActivityPanel = () => {
 		return null;
 	}
 
+	const getInitialState = ( panelId ) => {
```
Nit: how about `getInitialOpenState`, since this is just meant for the `initialOpen` prop and not component state.

octaedro replied:
Sounds good to me. I changed it in the commit da925e6

jeffstieler commented on packages/components/src/abbreviated-card/index.js:
```diff
@@ -0,0 +1,58 @@
+			) }
+		>
+			<CardBody size={ null }>
+				<Link href={ href } onClick={ onClick } type="wp-admin">
```
Should we be hardcoding these `Link` types to `wp-admin`? If we're on an Analytics page they could be `wc-admin` to avoid a full pageload.

octaedro replied:
Good suggestion! I modified this in the commit e4e7bdb and 36202c0

jeffstieler commented on client/header/activity-panel/panels/inbox/abbreviated-card.js:
```diff
@@ -0,0 +1,36 @@
+			</CardBody>
+		</Card>
+	);
+};
```
Is this file defunct?

octaedro replied:
Nice catch! I removed it in the commit 4e05beb

octaedro committed:
Added AbbreviatedNotificationsPanel

octaedro committed:
Added tests

octaedro committed:
Renamed `getUnreadNotes` to `isNotesPanelVisible`

octaedro committed:
Removed abbreviated-card.js

octaedro committed:
Added singular/plural copy handling

octaedro committed:
Renamed method `getInitialState` to `getInitialOpenState`

octaedro committed:
Fixed Link prop

octaedro committed:
Revert "Fixed Link prop"

This reverts commit 74e6a7fae030766eb5d6be098caa15478f2cb2c6.

octaedro committed:
Fixed Link prop

octaedro committed:
Added task list visibility check

octaedro committed:
Fixed scroll after redirect

# Conflicts:
#	client/task-list/index.js

octaedro committed:
Added propType to `AbbreviatedCard`

octaedro commented:
Hi @jeffstieler,
I applied the changes that you suggested. Could you take another look at this PR?

jeffstieler commented:
I gave this a spin to test the UX and it's looking good! (Nice find on the `<ScrollTo>` component)

I'll review the code in depth tomorrow morning.

What do you think about changing the logic in https://github.com/woocommerce/woocommerce-admin/blob/0677c60f5eaeb9b01d0fad864c0b13a96d74972b/docs/examples/extensions/add-task/woocommerce-admin-add-task-example.php#L15-L21

to be:

```php
if (
	! class_exists( 'Automattic\WooCommerce\Admin\Loader' ) ||
	! \Automattic\WooCommerce\Admin\Loader::is_admin_or_embed_page() ||
	! Onboarding::should_show_tasks()
) {
	return;
}
```

so the example plugin can be used to see the "Things to do next" in the inbox panel?

jeffstieler commented on client/header/activity-panel/unread-indicators.js:
```diff
@@ -57,3 +65,56 @@ export function getUnreadNotes( select ) {
+	).length;
+}
+
+export function isAbbreviatedPanelVisible(
```
If we switch to the prop drilling approach mentioned in my comment about the `applyFilters()` usage, I think we can remove this.

jeffstieler reviewed (changes requested):
This tested well and for the most part the code looks good. I'm wary of adding 2 more applications of the `woocommerce_admin_onboarding_task_list` filter - so let's discuss alternatives.

octaedro committed:
Fixed `Add-task doc example

octaedro committed:
Removed default values from ActivityPanel

octaedro committed:
Fixed multiple calls to a filter

octaedro commented on client/header/activity-panel/index.js:
```diff
@@ -66,26 +69,36 @@ export const ActivityPanel = ( { isEmbedded, query, userPreferencesData } ) => {
 	const {
-		hasUnreadNotes,
+		hasUnreadNotes = false,
+		hasAbbreviatedNotifications = false,
```
Good catch! I fixed it in the commit afd616e

octaedro commented on client/header/activity-panel/panels/inbox/abbreviated-notifications-panel.js:
```diff
@@ -0,0 +1,194 @@
+			'woocommerce_admin_onboarding_task_list',
+			[],
+			query
+		);
```
It's a good idea, I changed that in the commit 9df3aa0
I'm passing the value as a prop now and I will refactor the original usage of `woocommerce_admin_onboarding_task_list` in a follow-up PR.

octaedro commented:
Hi @jeffstieler,
Sorry for pinging you again, I made the changes you suggested.
Could you take a look at this PR again?

jeffstieler reviewed (approved):
Code looks good and tested well. :shipit: ❗
